yarn-project/simulator/src/avm/avm_machine_state.ts

```diff
@@ -80,6 +104,15 @@ export class AvmMachineState {
     this.output = output;
   }
 
+  /**
+   * Flag an exceptional halt. Clears gas left and sets the reverted flag. No output data.
+   */
+  protected exceptionalHalt() {
+    GasDimensions.forEach(dimension => (this[`${dimension}Left`] = 0));
```

Comment:
@fcarreiro @spalladino
What is the plan on how we're going to prove to the public kernel that I threw an exception due to "out of gas"?
In my mind, it would need to be something like "I had state S and was about to execute X". So it would provide S (the state before the operation) and X. In this case, we wouldn't want to zero out gas.

Reply:
Isn't this handled by the public VM? i.e. you don't need to prove it to the kernel, the VM says so.

Reply:
Got it. Even still, I'm not positive on what the outputs should be. Suppose I specify an l2 gas limit of 10, and i try to run two opcodes, one that consumes 1, and another that would have consumed 100. Do we charge the user 1 gas, or 10?

Reply:
Yep, I understand this is "just" the simulation side of things, the proof is produced by the circuit. And this clearing of gas is specified in the yellow paper.

Reply:
> Do we charge the user 1 gas, or 10?

I understand we always charge them all of their gas (in all dimensions!) if there is an out of gas.

Reply:
I don't think we would charge them the gas from all dimensions would we? e.g. I wouldn't lose all my L2 gas if I emitted a note hash right at the start of public execution that breached my DA gas limit.

Reply:
I think the YP should be taken as "malleable" in most things gas-related. It doesn't really change much about what the VM actually has to do, so I would defer to protocol experts (i.e., you). However, it might be worth checking if having to calculate the actual gas, etc, has any impact on the actual VM circuit. Maybe resetting all was chosen because of that? I'll bring Sean's attention to this.

Reply:
Yep, the vm circuit can catch an underflow on the gas left column and set a reverted flag.